lucene/core/src/java/org/apache/lucene/util/IntArrayDocIdSet.java

```diff
@@ -72,7 +72,12 @@ public int nextDoc() throws IOException {
 
     @Override
     public int advance(int target) throws IOException {
-      i = Arrays.binarySearch(docs, i + 1, length, target);
+      int bound = 1;
+      int offset = Math.max(0, i);
+      while(offset + bound < length && docs[offset + bound] < target) {
+        bound *= 2;
+      }
+      i = Arrays.binarySearch(docs, offset + bound / 2, Math.min(offset + bound, length), target);
```

Comment:
`bound/2` is generally the previous bound that we tested, except when `bound` is equal to 1. It won't break in that case since callers are not supposed to call advance on a target that is lte the current doc ID, but this might still make room for bugs?